### PR TITLE
[TwigBridge] Add row_attr to all form themes

### LINKED ISSUE #33552: [Form][TwigBridge] Add row_attr to all form themes
**Symfony version(s) affected**: 4.3+

**Description**  
In #30320 we added the `row_attr` option ... but only for the `form_div_layout.html.twig` theme.

We should add it too to all the other built-in themes: https://github.com/symfony/symfony/tree/4.3/src/Symfony/Bridge/Twig/Resources/views/Form

### PR DESCRIPTION
**Pull request by fancyweb**
| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #33552
| License       | MIT
| Doc PR        | -

The rules I applied:
- Always done on the first HTML tag of the row.
- Current existing row attrs (`class` or `style`) are applied unless they are defined by the `row_attr` override. They can be removed if they are explicitly set to `false`.

Starting from: 
```
<div class="form-group">
```

With `row_attr: {foo: "bar"}`:
```
<div foo="bar" class="form-group">
```

With `row_attr: {class: "ccc"}`:
```
<div class="ccc">
```

With `row_attr: {foo: "bar", class: false}`:
```
<div foo="bar">
```

### CODE REVIEW

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_horizontal_layout.html.twig [hunk 1]
```diff
@@ -27,7 +27,7 @@ col-sm-2
     {%- if help is not empty -%}
         {%- set widget_attr = {attr: {'aria-describedby': id ~"_help"}} -%}
     {%- endif -%}
-    <div class="form-group{% if (not compound or force_error|default(false)) and not valid %} has-error{% endif %}">
+    <div {% with {attr: row_attr.class is defined ? row_attr : row_attr|default({})|merge({class: 'form-group' ~ ((not compound or force_error|default(false)) and not valid ? ' has-error')})} %}{{ block('attributes') }}{% endwith %}>
```
Comment (stof) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_horizontal_layout.html.twig]: so this means that if I want to add a row_attr class (to apply a JS plugin for instance), I have to provide the classes normally part of the theme explicitly ? This makes the theme much harder to use. I suggest merging classes instead.

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_horizontal_layout.html.twig [hunk 2]
```diff
@@ -38,7 +38,7 @@ col-sm-2
 {%- endblock form_row %}
 
 {% block submit_row -%}
-    <div class="form-group">{#--#}
+    <div {% with {attr: row_attr.class is defined ? row_attr : row_attr|default({})|merge({class: 'form-group'})} %}{{ block('attributes') }}{% endwith %}>{#--#}
```
Comment (stof) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_horizontal_layout.html.twig]: `|default({})` is useless. `row_attr` is defined and is always an array, thanks to the FormType supporting it.

Comment (fancyweb) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_horizontal_layout.html.twig]: Where because I can't find it? The default is already there in the form_div_layout.

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_horizontal_layout.html.twig (hunk 1 above)
Comment (fancyweb) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_horizontal_layout.html.twig]: Yes. I thought about it but if we always merge defaults, then you cannot get rid of them if you want to. So it's "redefine everything if you wanna add one but you are able to remove the defaults" vs "merge but the defaults will always be here".

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_horizontal_layout.html.twig (hunk 2 above)
Comment (xabbuh) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_horizontal_layout.html.twig]: will be fixed in #33682

Comment (fancyweb) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_horizontal_layout.html.twig]: Unresolving because that means I must remove all `|default` here.

**Comment by xabbuh** [comment]
@fancyweb FYI, #33688 has been merged

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_horizontal_layout.html.twig (hunk 1 above)
Comment (fancyweb) [reply src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_horizontal_layout.html.twig]: @stof Just looked at this again and your suggestion is better for DX. Let's do the same thing that for the label attr -> merge classes but put the default classes in a new `form_row_class` block so it can be extended easily (to remove them for example). Does that work for you?

**Comment by fancyweb** [comment]
Here is what I did : 

1. Remove all defaults (cf https://github.com/symfony/symfony/pull/33688)

2. About https://github.com/symfony/symfony/pull/33573#discussion_r324256780 -> Always add current attr values (like what we do for the label attr class). That means end users can easily add attr values but not remove them. That's consistent with the current codebase + better for DX. If end users wants to remove default classes, they just have to override the whole block like currently. Creating dedicated blocks as I suggested is possible but requires more code and is probably not needed. Removing default classes makes no sense anyway (since it probably break the display).

File: src/Symfony/Component/Form/Tests/AbstractTableLayoutTest.php [hunk 3]
```diff
@@ -79,7 +79,7 @@ public function testRepeatedRow()
         /following-sibling::td
             [./input[@id="name_second"]]
     ]
-/following-sibling::tr[@style="display: none"]
+/following-sibling::tr[@style="display: none;"]
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Tests/AbstractTableLayoutTest.php]: can't we get rid of this change? I fear it might break cross testing - or do we need to bump `$supportedFeatureSetVersion` in child classes? @xabbuh wdyt?

Comment (fancyweb) [reply src/Symfony/Component/Form/Tests/AbstractTableLayoutTest.php]: Those changes are only necessary because I chose to prepend the `display: none` in https://github.com/symfony/symfony/pull/33573/files#diff-c66befd9821a00f467f17bc882749b85R30 to avoid issue if the added "style" by the end user  doesn't end with a `;`. I guess we can find a work around to append it instead and revert all impacted files.

Comment (fancyweb) [reply src/Symfony/Component/Form/Tests/AbstractTableLayoutTest.php]: I did it, look at https://github.com/symfony/symfony/pull/33573/files#diff-c66befd9821a00f467f17bc882749b85R30.

File: src/Symfony/Bridge/Twig/composer.json [hunk 4]
```diff
@@ -26,7 +26,7 @@
         "symfony/asset": "~3.4|~4.0",
         "symfony/dependency-injection": "~3.4|~4.0",
         "symfony/finder": "~3.4|~4.0",
-        "symfony/form": "^4.3.4",
+        "symfony/form": "^4.3.5",
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Twig/composer.json]: can be reverted (same below)

Comment (fancyweb) [reply src/Symfony/Bridge/Twig/composer.json]: 4.3.4 doesn't have https://github.com/symfony/symfony/pull/33688 so `row_attr` can not exist. I guess the bump + `|default({})` removal in `form_div_layout` should have been made there actually.

**Comment by loevgaard** [comment]
What's the status with this PR? :)

File: src/Symfony/Bridge/Twig/composer.json [hunk 5]
```diff
@@ -26,7 +26,7 @@
         "symfony/asset": "~3.4|~4.0",
         "symfony/dependency-injection": "~3.4|~4.0",
         "symfony/finder": "~3.4|~4.0",
-        "symfony/form": "^4.3.4",
+        "symfony/form": "^4.3.9",
```
Comment (xabbuh) [inline src/Symfony/Bridge/Twig/composer.json]: Why is that needed?

File: src/Symfony/Bundle/FrameworkBundle/composer.json [hunk 6]
```diff
@@ -38,7 +38,7 @@
         "symfony/css-selector": "~3.4|~4.0",
         "symfony/dom-crawler": "^4.3",
         "symfony/polyfill-intl-icu": "~1.0",
-        "symfony/form": "^4.3.4",
+        "symfony/form": "^4.3.9",
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/composer.json]: Why is that needed?

File: src/Symfony/Bridge/Twig/composer.json (hunk 5 above)
Comment (fancyweb) [reply src/Symfony/Bridge/Twig/composer.json]: My bad, 4.3.5 is enough, see https://github.com/symfony/symfony/pull/33573#discussion_r329006002

**Review by xabbuh — APPROVED** [review]

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @fancyweb.
